x-pack/plugins/ml/server/lib/capabilities/check_capabilities.ts

```diff
@@ -47,3 +50,27 @@ function disableAdminPrivileges(capabilities: MlCapabilities) {
   capabilities.canCreateAnnotation = false;
   capabilities.canDeleteAnnotation = false;
 }
+
+export type HasMlCapabilities = (capabilities: MlCapabilitiesKey[]) => Promise<void>;
+
+export function hasMlCapabilitiesProvider(resolveMlCapabilities: ResolveMlCapabilities) {
+  return (request: KibanaRequest): HasMlCapabilities => {
+    let mlCapabilities: MlCapabilities | null = null;
+    return async (capabilities: MlCapabilitiesKey[]) => {
+      try {
+        mlCapabilities = await resolveMlCapabilities(request);
+      } catch (e) {
+        mlLog.warn('Unable to perform ML capabilities check');
+        throw Error(e);
+      }
+
+      if (mlCapabilities === null) {
+        throw Error('ML capabilities have not been initialized');
+      }
+
+      if (capabilities.every((c) => mlCapabilities![c] === true) === false) {
+        throw Error('Insufficient privileges to access feature');
```

Comment:
I was wondering if we could use a custom `Error` subclass here to enable the routes to translate it into a 403 instead of 500 error. :thinking: 

We're doing something like that in the `infra` plugin here:

https://github.com/elastic/kibana/blob/e8cf08fb3e640d6f15a3056556783af445dfa70c/x-pack/plugins/infra/server/lib/log_analysis/errors.ts#L16-L21

Alternatively, a custom marker attribute on the `Error` subclass could also be used for identification.

What do you think?

Reply:
yes, I like this idea.
added in fbac69ffbff0a6912045088fe94bd817c4547c7b

Reply:
awesome, thank you :heart: